### PR TITLE
Add binary heap implementation.

### PR DESCRIPTION
**Pull request by Walt280**
Add an implementation of binary heap. I feel this would be beneficial as a easy-to-understand introduction to heaps.

- [X] I have performed a self-review of my code
- [X] My code follows the style guidelines of this project
- [X] I have added comments to hard-to-understand areas of my code
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] I have made corresponding changes to the comments
- [X] I have made corresponding changes to the README.md

### CODE REVIEW

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 5
           

Complexity increasing per file
==============================
- DataStructures.Tests/BinaryHeapTests.cs  2
- DataStructures/BinaryHeap.cs  11
         

Clones added
============
- DataStructures.Tests/BinaryHeapTests.cs  27
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6375083&bid=20683860)

File: DataStructures/BinaryHeap.cs [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+                HeapifyDown(left);
+            }
+            else if (right < Count && rightLargerThanElem)
+            {
+                Swap(elemIdx, right);
+                HeapifyDown(right);
+            }
+            else
+            {
+                return;
```
Comment (siriak) [inline DataStructures/BinaryHeap.cs:260]: Codacy found an issue: [Remove this redundant jump.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6375083)

File: DataStructures/BinaryHeap.cs [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+
+        /// <summary>
+        /// Recursive function to restore heap properties.
+        /// </summary>
+        /// <remarks>
+        /// Restores heap property by swapping the element at <paramref name="elemIdx"/>
+        /// with the larger of its children.
+        /// </remarks>
+        /// <param name="elemIdx">The element to check with its children.</param>
+        private void HeapifyDown(int elemIdx)
```
Comment (siriak) [inline DataStructures/BinaryHeap.cs:225]: Codacy found an issue: [The Cyclomatic Complexity of this method is 11 which is greater than 10 authorized.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6375083)

File: DataStructures.Tests/BinaryHeapTests.cs [hunk 3]
```diff
@@ -0,0 +1,182 @@
+﻿using System;
+using System.Collections.Generic;
+
+using NUnit.Framework;
+
+namespace DataStructures.Tests
+{
+    class BinaryHeapTests
```
Comment (siriak) [inline DataStructures.Tests/BinaryHeapTests.cs:8]: Codacy found an issue: [Add a 'protected' constructor or the 'static' keyword to the class declaration.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6375083)

File: DataStructures/BinaryHeap.cs [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+        {
+            var left = (2 * elemIdx) + 1;
+            var right = (2 * elemIdx) + 2;
+
+            bool leftLargerThanElem = left < Count ? comparer.Compare(data[elemIdx], data[left]) < 0 : false;
```
Comment (siriak) [inline DataStructures/BinaryHeap.cs:230]: Codacy found an issue: [Remove the unnecessary Boolean literal(s).](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6375083)

File: DataStructures/BinaryHeap.cs [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+            bool rightLargerThanElem = right < Count ? comparer.Compare(data[elemIdx], data[left]) < 0 : false;
```
Comment (siriak) [inline DataStructures/BinaryHeap.cs:231]: Codacy found an issue: [Remove the unnecessary Boolean literal(s).](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6375083)

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- DataStructures.Tests/BinaryHeapTests.cs  2
- DataStructures/BinaryHeap.cs  9
         

Clones added
============
- DataStructures.Tests/BinaryHeapTests.cs  27
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6375083&bid=20683860)

File: DataStructures/BinaryHeap.cs [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+            var leftLargerThanElem = left < Count && comparer.Compare(data[elemIdx], data[left]) < 0;
+            var rightLargerThanElem = right < Count && comparer.Compare(data[elemIdx], data[right]) < 0;
+            var leftLargerThanRight = left < Count && right < Count && comparer.Compare(data[left], data[right]) > 0;
+
+            if (leftLargerThanElem && leftLargerThanRight)
+            {
+                Swap(elemIdx, left);
+                HeapifyDown(left);
+            }
+            else if (rightLargerThanElem && !leftLargerThanRight)
```
Comment (siriak) [inline DataStructures/BinaryHeap.cs:239]: Codacy found an issue: [Add the missing 'else' clause.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6375083)

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- DataStructures.Tests/BinaryHeapTests.cs  2
- DataStructures/BinaryHeap.cs  9
         

Clones added
============
- DataStructures.Tests/BinaryHeapTests.cs  27
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6375083&bid=20683860)

File: DataStructures/BinaryHeap.cs [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+            {
+                Swap(elemIdx, right);
+                HeapifyDown(right);
+            }
+
+            // if left and right child values compare smaller than the current value, the heap property has been restored.
+            // no need to do anything else.
+            else
+            {
```
Comment (siriak) [inline DataStructures/BinaryHeap.cs:248]: Codacy found an issue: [Either remove or fill this block of code.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6375083)

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- DataStructures.Tests/BinaryHeapTests.cs  2
- DataStructures/BinaryHeap.cs  9
         

Clones added
============
- DataStructures.Tests/BinaryHeapTests.cs  27
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6375083&bid=20683860)

File: DataStructures/BinaryHeap.cs (hunk 7 above)
Comment (siriak) [inline DataStructures/BinaryHeap.cs:248]: Codacy found an issue: [Either remove or fill this block of code.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6375083)

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- DataStructures.Tests/BinaryHeapTests.cs  2
- DataStructures/BinaryHeap.cs  9
         

Clones added
============
- DataStructures.Tests/BinaryHeapTests.cs  27
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6375083&bid=20683860)

File: DataStructures.Tests/BinaryHeapTests.cs [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+using NUnit.Framework;
+
+namespace DataStructures.Tests
+{
+    static class BinaryHeapTests
+    {
+        [Test]
+        public static void Constructor_UseCustomComparer_BuildCorrectHeap()
+        {
+            var revHeap = new BinaryHeap<int>(Comparer<int>.Create((x, y) => x.CompareTo(y) * -1));
```
Comment (siriak) [inline DataStructures.Tests/BinaryHeapTests.cs:13]:
```suggestion
            var revHeap = new BinaryHeap<int>(Comparer<int>.Create((x, y) => y.CompareTo(x)));
```
Isn't it the same?

File: DataStructures.Tests/BinaryHeapTests.cs [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+        {
+            var heap = new BinaryHeap<int>();
+            var elems = new[] { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 };
+            foreach (int i in elems)
+            {
+                heap.Push(i);
+            }
+
+            Assert.Throws<ArgumentException>(() => heap.Push(1));
+        }
```
Comment (siriak) [inline DataStructures.Tests/BinaryHeapTests.cs:50]: Why? Heap can contain equal elements

File: DataStructures/BinaryHeap.cs [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+using System.Collections.Generic;
+
+namespace DataStructures
+{
+    /// <summary>
+    /// A generic implementation of a binary heap.
+    /// </summary>
+    /// <remarks>
+    /// A binary heap is a complete binary tree that satisfies the heap property;
+    /// that is every node in the tree compares greater/less than its left and right
```
Comment (siriak) [inline DataStructures/BinaryHeap.cs:11]: From Wikipedia

> Heap property: the key stored in each node is either **greater than or equal to** (≥) or **less than or equal to** (≤) the keys in the node's children, according to some total order.

So heap can contain the same element multiple times.

File: DataStructures/BinaryHeap.cs [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+            T elem = data[0];
+            data[0] = data[data.Count - 1];
+            data.RemoveAt(data.Count - 1);
+            HeapifyDown(0);
+
+            return elem;
+        }
+
+        /// <summary>
+        /// Return the top/root of the heap without remove it.
```
Comment (siriak) [inline DataStructures/BinaryHeap.cs:96]:
```suggestion
        /// Return the top/root of the heap without removing it.
```

File: DataStructures/BinaryHeap.cs [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+        /// </summary>
+        /// <remarks>
+        /// Removing from the heap is done by swapping the top/root with the last element in
+        /// the backing list, removing the last element, and pushing the new root down
+        /// until the heap property is restored.
+        /// </remarks>
```
Comment (siriak) [inline DataStructures/BinaryHeap.cs:102]: Is this relevant?

File: DataStructures/BinaryHeap.cs [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+        /// <returns>true if element is in the heap, false otherwise.</returns>
+        public bool Contains(T element) => data.Contains(element);
+
+        /// <summary>
+        /// Remove an element from the heap.
+        /// </summary>
+        /// <remarks>
+        /// In removing an element from anywhere in the heap, we only need to push down or up
+        /// the replacement value depending on how the removed value compares to its
+        /// replacment value.
```
Comment (siriak) [inline DataStructures/BinaryHeap.cs:154]:
```suggestion
        /// replacement value.
```

File: DataStructures/BinaryHeap.cs [hunk 14]
```diff
@@ -0,0 +1,251 @@
...
+                    if (comparer.Compare(tmp, data[idx]) > 0)
+                    {
+                        HeapifyDown(idx);
+                    }
+                    else
+                    {
+                        HeapifyUp(idx);
+                    }
+                }
+            }
```
Comment (siriak) [inline DataStructures/BinaryHeap.cs:183]:
```suggestion
                Swap(idx, data.Count - 1);
                T tmp = data[data.Count - 1];
                data.RemoveAt(data.Count - 1);

                if (idx < data.Count)
                {
                    if (comparer.Compare(tmp, data[idx]) > 0)
                    {
                        HeapifyDown(idx);
                    }
                    else
                    {
                        HeapifyUp(idx);
                    }
                }
```
Else clause not needed, code won't execute any further after throw

**Review by siriak — COMMENTED** [review]
Looks good, I've left some comments, mostly small stuff like typos, and an important in my opinion point that binary heap can contain equal elements.

File: DataStructures/BinaryHeap.cs (hunk 12 above)
Comment (Walt280) [reply DataStructures/BinaryHeap.cs]: That section was in the wrong place; I have moved it above the correct function.

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- DataStructures.Tests/BinaryHeapTests.cs  2
- DataStructures/BinaryHeap.cs  9
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6375083&bid=20683860)

**Review by siriak — APPROVED** [review]
Thanks, looks good!

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- DataStructures.Tests/BinaryHeapTests.cs  2
- DataStructures/BinaryHeap.cs  9
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6375083&bid=20683860)

**Pull request merged**